Game OSD: Show help dialog on first view

garbear opened the pull request:
## Description
This adds a help dialog that is shown the first time a game is played.

The help dialog shows the button combo to activate the menu and a picture of the controller the button combo belongs to.

The problem is we don't know the layout of the controller in the user's hand. For now, we show a SNES controller because that is the layout used by the "retropad" abstraction in the libretro API.

## Motivation and Context
Discussion: https://forum.kodi.tv/showthread.php?tid=335300&pid=2766690#pid2766690

Motivation: https://github.com/xbmc/xbmc/pull/14398

## How Has This Been Tested?
Tested on OSX.

Test builds: https://github.com/garbear/xbmc/releases

## Screenshots (if appropriate):
Shown when FullscreenGame window is opened for the first time:

<img width="1920" alt="game osd help" src="https://user-images.githubusercontent.com/531482/45601520-e87b7880-b9c2-11e8-8ef9-1cc71f1bbb0d.png">


Shown subsequently:

![screen shot 2018-09-11 at 2 14 04 pm](https://user-images.githubusercontent.com/531482/45389109-d919bb00-b5cf-11e8-8c09-2fd36e1253a9.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

garbear committed:
CStaticListProvider: Remove redundant check

garbear committed:
Game OSD: Show help dialog on first view

This adds a help dialog to the game OSD and encapsulates the GUI playback
control mechanisms.

The volume dialog has been modified to play the game in the background
for audible feedback.

da-anda commented:
wouldn't a little description/text be nice instead of just the gamepad + one menu line? Not sure everybody will get what this is supposed to mean. Maybe something along the lines
`During gameplay all button and key presses of your input devices are forwarded to the game for the best gaming experience. In order to pause the game or to exit this game mode, press this combination`.
In addition to that, we could show a subtile temporary overlay like `Game mode ON: press SELECT+X to exit` for like 3 seconds (like a Kaitoast) every time the user enters fullscreen, because even though you might remember it the first day after reading the "first time dialog", you might have forgotten the combination one week later.

garbear commented:
You are right, my help dialog is confusing. I'll add some basic text.

I want absolute minimum information exchange to happen. The user does not want to spend time learning that could be spent playing. I want the user to know one thing: the combo for the menu. Once in the menu, they can see the available actions, but they need to know how to get in.

Rechi commented on xbmc/games/dialogs/osd/DialogGameOSD.cpp:
```diff
@@ -7,14 +7,54 @@
 using namespace KODI;
 using namespace GAME;
 
+#define CONTROL_OSD_GAME_CONTROLLER   1101
```
`static const int CONTROL_OSD_GAME_CONTROLLER = 1101` or `static constexpr int CONTROL_OSD_GAME_CONTROLLER = 1101`

da-anda commented:
I don't think that showing something like a Kaitoast displaying the key-combo on how to exit game mode when you enter fullscreen mode will distract anyone from playing the game, especially because many games/roms show some "intro" animation first anyways. Also, we can be smart about it and only show it on game startup and not while game is running and people back out and in etc.
I'm thinking about occasional retro gamers and multi-user house holds that might share the same KODI profile. Since the big dialog of this PR is only shown once, a second users that might give retro gaming a try some other day would not know anything about this dialog and be trapped. A Kaitoast on game start might help. But that's just my POV

garbear commented:
So a toast dialog saying "Press Select + X to pause"? That's an alternative, but I'm really not a fan of annoying the user with dialogs every time a game starts. With my large intrusive dialog, they see it once, they tell their friends, and if anyone is ever stuck a factory reset or new profile will bring up the dialog again.

da-anda commented:
I'd personally give a more detailed explanation in this one time dialog. "Press Select + X to open the menu" IMO still doesn't help unexperienced users. They will ask "What menu, and what is it good for?". So I'd explain that KODI is entering what I'd probably call "game mode" which will forward any key or button presses to the game. In order to exit this mode or pause the game one has to press Select+X. It's really not that much to read, but IMO will help unexperienced users a lot and give them a reason they understand why they should remember this key-combo. But up to you.

garbear commented:
> They will ask "What menu, and what is it good for?"

Yes they will. Your Jedi mind powers can see well into the minds of the technopleebs.

I'm against an explicit "game mode" - modes add mental overhead because you gotta keep track of them. I think game mode should be implicit - if you see no GUI, you're playing the game. If you see GUI, the game is paused and buttons control the menus you see.

Next iteration: what are your thoughts on this?

<img width="1920" alt="screen shot 2018-09-14 at 11 59 12 am" src="https://user-images.githubusercontent.com/531482/45570234-c8a75180-b816-11e8-9d06-5ca444d291d0.png">

da-anda commented:
better, thanks. I still think that it would be good to let people know that none of the default Kodi shortcuts (not even back/esc) will work while playing a game so that there is no surprise, but up to you.

I just had a thought, because I somehow had to think about how in-game tutorials work. There you get a description on which button combo to press in order to perform a certain action and you can only proceed with the game when you managed to get the combo right. So what do you think about the user having to press exactly this key combo to close the dialog? That way he already has to figure out which buttons they are? We'd have to change the text to address this though. Something like:
`You can press SELECT+X at any time to pause the game and open the in-game menu. Open it now and select "Resume" in order to proceed`.

PS: sorry for kinda "blocking" this PR with my ideas. Feel free to ignore them :)

garbear commented:
No, I appreciate your ideas. I like your idea of blocking the user until they get the right combo, but this doesn't work if the user doesn't have a controller. I'm trying to cater to the lowest common denominator, which is why I think remotes and keyboards should still be able to dismiss the dialog.

One change of yours I'm considering is the "tutorial" aspect - when the user presses the combo, instead of resuming gameplay, show the real menu so they get a cause-effect type of experience.

garbear commented:
I modified the help dialog to bring up the in-game menu immediately upon dismissal for a "tutorial"-like feel, aiding in reinforcement. Good idea.

However, I had to refactor, because the help dialog was controlled almost entirely by the skin. I moved some logic into core and encapsulated it in a new class.

@da-anda Any other changes you think we need to make before release?

da-anda commented:
I suppose we can't detect if the user is using a controller when launching the game or not? If so, it would ofc be nice if this dialog (which, if I understood you correctly is only relevant for controllers) would only show up when a controller has been used to launch the game, or when first used in fullscreen game mode (first button press or analog axis movement).

If this is not possible, I'd probably adjust the text a bit so that users that play casual games with a remote won't be confused by the controller picture in that dialog. Maybe something like:

`When you play with a controller, you can press Select + X to open the in-game menu at any time. Proceed to get familiar with the in-game menu now`. 

Alternatively, if we know which keymap/button combination will trigger this menu when using other input devices (ieyboard / remote), I'd probably write the following:
`Press the menu button (or Select + X on a controller) to open the in-game menu at any time. Proceed to get familiar with the in-game menu now`. 

Feel free to adjust the wording, since I'm no native speaker and it might sound odd to you :)

garbear commented:
Cool, I think we're on the same page as to how we handle controller vs. keyboard vs. remote. Either only show the help when a controller is attached, or adjust the text to mesh with the device the user is holding.

Controller and device detection is a v19 feature though, so I think I'll stick with the current dialog for v18, as limited as it is. I'll upgrade it with these ideas when the device-detection feature is implemented.

Gonna test a bit more then merge tonight or tmrw.

garbear committed:
Fix crash when checking if service manager is up

The crash occurs because the service manager is not alive to check if it
is up.

garbear commented:
The code needed more work, but stuck in a plane for 9 hours is a boon for productivity, so it's ready now. In this new version, #14438 is not needed.

da-anda commented:
@garbear which action triggers the game osd on remotes and the keyboard?
Regular Menu action? (on mobile ATM, so can't check the code). If so, what
about my suggested text change from above?
Other than that, PR is fine from my end. Thanks for this usability
improvement.

Garrett Brown <notifications@github.com> schrieb am Do., 20. Sep. 2018,
16:44:

> The code needed more work, but stuck in a plane for 9 hours is a boon for
> productivity, so it's ready now. In this new version, #14438
> <https://github.com/xbmc/xbmc/pull/14438> is not needed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/14425#issuecomment-423210293>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABGwUVYf1xNBE_Kerxrwj98Z5w5uCW_Pks5uc6nagaJpZM4WkV9k>
> .
>

da-anda commented:
tested retroplayer using the remote on my SHIELD yesterday, and it appears as if a simple "select" action (-> ENTER on the keyboard) would already trigger the in-game menu. Is this desired? If we would remap this to the "menu" action, we could use `Press the menu button (or Select + X on a controller) to open the in-game menu at any time. Proceed to get familiar with the in-game menu now.` as wording for the dialog.

garbear commented:
I'll use your ideas in our next version. keyboard isn't supported in joystick games right now, so I didn't care what keys bring up the osd. gotta start small
